src/ast/passes/semantic_analyser.cpp

```diff
@@ -724,46 +724,48 @@ void SemanticAnalyser::visit(Call &call)
     call.type = CreateStats(true);
   } else if (call.func == "delete") {
     check_assignment(call, false, false, false);
-    if (check_varargs(call, 1, std::numeric_limits<size_t>::max())) {
-      if (call.vargs.size() == 2) {
-        if (!call.vargs.at(0)->is_map)
-          LOG(ERROR, call.vargs.at(0)->loc, err_) << DELETE_ERROR;
-
-        auto *key_arg = call.vargs.at(1);
-        if (!key_arg->is_map) {
-          Map &map = static_cast<Map &>(*call.vargs.at(0));
+    if (check_varargs(call, 1, 2)) {
+      if (!call.vargs.at(0)->is_map) {
+        LOG(ERROR, call.vargs.at(0)->loc, err_) << DELETE_ERROR;
+      } else {
+        Map &map = static_cast<Map &>(*call.vargs.at(0));
+        if (call.vargs.size() == 1) {
+          if (map.key_expr) {
+            // We're modifying the AST here to support the deprecated delete
+            // API. Once we remove the old API, we can delete this.
+            call.vargs.push_back(map.key_expr);
+            map.key_expr = nullptr;
+          } else if (is_final_pass()) {
+            auto *map_key_type = get_map_key_type(map);
+            if (map_key_type && !map_key_type->IsNoneTy()) {
+              LOG(ERROR, call.vargs.at(0)->loc, err_) << DELETE_ERROR;
+            }
+          }
+        } else {
           if (map.key_expr) {
             LOG(ERROR, call.vargs.at(0)->loc, err_)
                 << "delete() expects a map with no keys for the first argument";
-          } else {
-            auto *mapkey = get_map_key_type(map);
-            if (mapkey) {
-              auto key_type = create_key_type(call.vargs.at(1)->type,
-                                              call.vargs.at(1)->loc);
-              if (!mapkey->IsSameType(key_type) ||
-                  !key_type.FitsInto(*mapkey)) {
-                LOG(ERROR, call.vargs.at(1)->loc, err_)
-                    << "Argument mismatch for " << map.ident << ": "
-                    << "trying to delete with key of type: '"
-                    << call.vargs.at(1)->type << "' when map has key of type: '"
-                    << *mapkey << "'";
+          }
+          auto *map_key_type = get_map_key_type(map);
+          if (map_key_type) {
+            auto &arg1 = *call.vargs.at(1);
+            if (is_final_pass() && arg1.is_map) {
+              Map &map_as_key = static_cast<Map &>(arg1);
+              auto *map_as_key_key_type = get_map_key_type(map_as_key);
+              if (!map_as_key.key_expr && map_as_key_key_type &&
+                  !map_as_key_key_type->IsNoneTy()) {
+                LOG(ERROR, map_as_key.loc, err_)
+                    << "Argument mismatch for " << map_as_key.ident << ": "
+                    << "trying to access with no arguments when map expects "
+                       "arguments: "
+                       "'"
+                    << *map_as_key_key_type << "'";
```

Comment:
It looks like we're duplicating some code from `visit(Map &map)` here. It might be cleaner to keep all key validation logic in that function.

The issue seems to be that we're setting `map.skip_key_validation` for delete when we don't actually want to skip validation for the second argument. What about changing the call to `skip_key_validation(const Call &call)` to only run on the first argument? https://github.com/bpftrace/bpftrace/blob/84cfa952f5c4d34e6b7771fdf42e1b7dd422b011/src/ast/passes/semantic_analyser.cpp#L592

Reply:
Good call. This made me realize I also need to apply this same logic for `has_key`.